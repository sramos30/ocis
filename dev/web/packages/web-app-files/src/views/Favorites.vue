<template>
  <div>
    <app-bar />
    <app-loading-spinner v-if="loadResourcesTask.isRunning" />
    <template v-else>
      <no-content-message v-if="isEmpty" id="files-favorites-empty" class="files-empty" icon="star">
        <template #message>
          <span v-translate>There are no resources marked as favorite</span>
        </template>
      </no-content-message>
      <resource-table
        v-else
        id="files-favorites-table"
        v-model="selectedResources"
        class="files-table"
        :class="{ 'files-table-squashed': !sidebarClosed }"
        :are-paths-displayed="true"
        :are-thumbnails-displayed="displayThumbnails"
        :resources="paginatedResources"
        :target-route="resourceTargetLocation"
        :header-position="fileListHeaderY"
        :sort-by="sortBy"
        :sort-dir="sortDir"
        @fileClick="$_fileActions_triggerDefaultAction"
        @rowMounted="rowMounted"
        @sort="handleSort"
      >
        <template #quickActions="props">
          <quick-actions class="oc-visible@s" :item="props.resource" :actions="app.quickActions" />
        </template>
        <template #contextMenu="{ resource }">
          <context-actions v-if="isResourceInSelection(resource)" :items="selectedResources" />
        </template>
        <template #footer>
          <pagination :pages="paginationPages" :current-page="paginationPage" />
          <list-info
            v-if="paginatedResources.length > 0"
            class="oc-width-1-1 oc-my-s"
            :files="totalFilesCount.files"
            :folders="totalFilesCount.folders"
            :size="totalFilesSize"
          />
        </template>
      </resource-table>
    </template>
  </div>
</template>

<script lang="ts">
import { mapGetters, mapState, mapActions } from 'vuex'
import ResourceTable from '../components/FilesList/ResourceTable.vue'

import FileActions from '../mixins/fileActions'
import MixinFilesListFilter from '../mixins/filesListFilter'
import MixinMountSideBar from '../mixins/sidebar/mountSideBar'
import { VisibilityObserver } from 'web-pkg/src/observer'
import { ImageDimension, ImageType } from '../constants'
import debounce from 'lodash-es/debounce'

import AppBar from '../components/AppBar/AppBar.vue'
import QuickActions from '../components/FilesList/QuickActions.vue'
import AppLoadingSpinner from 'web-pkg/src/components/AppLoadingSpinner.vue'
import NoContentMessage from 'web-pkg/src/components/NoContentMessage.vue'
import ListInfo from '../components/FilesList/ListInfo.vue'
import Pagination from '../components/FilesList/Pagination.vue'
import ContextActions from '../components/FilesList/ContextActions.vue'
import { createLocationSpaces } from '../router'
import { useResourcesViewDefaults } from '../composables'
import { defineComponent } from '@vue/composition-api'
import { Resource } from '../helpers/resource'

const visibilityObserver = new VisibilityObserver()

export default defineComponent({
  components: {
    AppBar,
    ResourceTable,
    QuickActions,
    AppLoadingSpinner,
    Pagination,
    NoContentMessage,
    ListInfo,
    ContextActions
  },

  mixins: [FileActions, MixinMountSideBar, MixinFilesListFilter],

  setup() {
    return {
      ...useResourcesViewDefaults<Resource, any, any[]>(),
      resourceTargetLocation: createLocationSpaces('files-spaces-personal-home')
    }
  },

  computed: {
    ...mapState(['app']),
    ...mapState('Files', ['files']),
    ...mapGetters('Files', ['highlightedFile', 'totalFilesCount', 'totalFilesSize']),
    ...mapState('Files/sidebar', { sidebarClosed: 'closed' }),
    ...mapGetters(['user', 'configuration']),

    isEmpty() {
      return this.paginatedResources.length < 1
    },

    displayThumbnails() {
      return !this.configuration.options.disablePreviews
    }
  },

  created() {
    this.loadResourcesTask.perform()
  },

  beforeDestroy() {
    visibilityObserver.disconnect()
  },

  methods: {
    ...mapActions('Files', ['loadPreview']),

    rowMounted(resource, component) {
      if (!this.displayThumbnails) {
        return
      }

      const debounced = debounce(({ unobserve }) => {
        unobserve()
        this.loadPreview({
          resource,
          isPublic: false,
          dimensions: ImageDimension.Thumbnail,
          type: ImageType.Thumbnail
        })
      }, 250)

      visibilityObserver.observe(component.$el, { onEnter: debounced, onExit: debounced.cancel })
    }
  }
})
</script>
