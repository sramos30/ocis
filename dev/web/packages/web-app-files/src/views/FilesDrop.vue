<template>
  <div id="files-drop-container" class="oc-height-1-1 oc-flex oc-flex-column oc-flex-between">
    <h1 class="oc-invisible-sr">{{ pageTitle }}</h1>
    <div class="oc-p oc-height-1-1">
      <div v-if="loading" key="loading-drop" class="oc-flex oc-flex-column oc-flex-middle">
        <h2 class="oc-login-card-title">
          <translate>Loading public linkâ€¦</translate>
        </h2>
        <oc-spinner :aria-hidden="true" />
      </div>
      <div v-else key="loaded-drop" class="oc-flex oc-flex-column oc-flex-middle oc-height-1-1">
        <div class="oc-text-center oc-width-1-1 oc-width-xxlarge@m">
          <h2 v-text="title" />
          <vue-dropzone
            id="oc-dropzone"
            :options="dropzoneOptions"
            :use-custom-slot="true"
            :include-styling="false"
            @vdropzone-file-added="dropZoneFileAdded"
          >
            <div class="oc-flex oc-flex-middle oc-flex-center oc-files-drop-drag-area">
              <oc-icon name="file-upload" />
              <translate>Drop files here to upload or click to select file</translate>
            </div>
          </vue-dropzone>
          <div id="previews" hidden />
        </div>
        <div
          class="oc-flex oc-flex-center oc-overflow-auto oc-width-1-1 oc-mt"
          :class="{ 'files-empty': !getUploadedFiles }"
        >
          <oc-table-simple v-if="getUploadedFiles" class="oc-width-1-1 oc-width-xxlarge@m">
            <oc-tbody>
              <oc-tr v-for="(file, key) in getUploadedFiles" :key="key">
                <oc-td class="oc-pl-rm" v-text="file.name" />
                <oc-td width="shrink" class="oc-text-nowrap oc-text-muted">
                  <oc-resource-size :size="file.size" />
                </oc-td>
                <oc-td width="shrink" class="oc-pr-rm">
                  <oc-icon
                    v-if="file.status === 'done'"
                    name="checkbox-circle"
                    size="small"
                    variation="success"
                  />
                  <oc-icon
                    v-if="file.status === 'error'"
                    name="information"
                    size="small"
                    variation="danger"
                  />
                  <oc-spinner
                    v-if="file.status === 'uploading' || file.status === 'init'"
                    size="small"
                    :aria-label="$_ocUploadingFileMessage(file.name)"
                  />
                </oc-td>
              </oc-tr>
            </oc-tbody>
          </oc-table-simple>
        </div>
        <div v-if="errorMessage" class="oc-text-center">
          <h2>
            <translate>An error occurred while loading the public link</translate>
          </h2>
          <p class="oc-m-rm" v-text="errorMessage" />
        </div>
      </div>
    </div>
    <div class="oc-text-center">
      <p v-text="configuration.currentTheme.general.slogan" />
    </div>
  </div>
</template>

<script>
import vue2DropZone from 'vue2-dropzone'
import { mapGetters } from 'vuex'
import Mixins from '../mixins.js'
import { DavProperties, DavProperty } from 'web-pkg/src/constants'
import { linkRoleUploaderFolder } from '../helpers/share'
import { createLocationOperations, createLocationPublic } from '../router'

export default {
  components: {
    vueDropzone: vue2DropZone
  },
  mixins: [Mixins],
  data() {
    return {
      loading: true,
      errorMessage: null,
      uploadedFiles: new Map(),
      uploadedFilesChangeTracker: 0
    }
  },
  computed: {
    ...mapGetters(['configuration']),
    ...mapGetters('Files', ['publicLinkPassword']),
    pageTitle() {
      return this.$gettext(this.$route.meta.title)
    },
    publicLinkToken() {
      return this.$route.params.token
    },
    title() {
      // share might not be loaded
      if (this.share) {
        const translated = this.$gettext('%{owner} shared this folder with you for uploading')
        return this.$gettextInterpolate(
          translated,
          { owner: this.share.getProperty(this.$client.publicFiles.PUBLIC_LINK_SHARE_OWNER) },
          true
        )
      }
      return ''
    },
    url() {
      return this.$client.publicFiles.getFileUrl(this.publicLinkToken) + '/'
    },
    getUploadedFiles() {
      return this.uploadedFilesChangeTracker && this.uploadedFiles.values()
    },
    dropzoneOptions() {
      return {
        url: this.url,
        clickable: true,
        createImageThumbnails: false,
        autoQueue: false,
        previewsContainer: '#previews'
      }
    }
  },
  mounted() {
    this.resolvePublicLink()
  },
  methods: {
    resolvePublicLink() {
      this.loading = true

      this.$client.publicFiles
        .list(this.publicLinkToken, this.publicLinkPassword, DavProperties.PublicLink, '0')
        .then((files) => {
          // Redirect to files list if the link doesn't have role "uploader"
          const sharePermissions = parseInt(files[0].getProperty(DavProperty.PublicLinkPermission))
          if (linkRoleUploaderFolder.bitmask(false) !== sharePermissions) {
            this.$router.replace(
              createLocationPublic('files-public-files', {
                params: { item: this.publicLinkToken }
              })
            )
            return
          }
          this.share = files[0]
        })
        .catch((error) => {
          // likely missing password, redirect to public link password prompt
          if (error.statusCode === 401) {
            this.$router.push(
              createLocationOperations('files-operations-resolver-public-link', {
                params: {
                  token: this.publicLinkToken
                }
              })
            )
            return
          }
          this.errorMessage = error
        })
        .finally(() => {
          this.loading = false
        })
    },
    dropZoneFileAdded(event) {
      const uploadId = event.upload.uuid
      this.uploadedFilesChangeTracker++
      this.uploadedFiles.set(uploadId, { name: event.name, size: event.size, status: 'init' })

      this.uploadQueue
        .add(() =>
          this.$client.publicFiles.putFileContents(
            this.publicLinkToken,
            event.name,
            this.publicLinkPassword,
            event,
            {
              // automatically rename in case of duplicates
              headers: { 'OC-Autorename': 1 },
              onProgress: (progressEvent) => {
                this.uploadedFilesChangeTracker++
                this.uploadedFiles.set(uploadId, {
                  name: event.name,
                  size: event.size,
                  status: 'uploading'
                })
              }
            }
          )
        )
        .then((e) => {
          this.uploadedFilesChangeTracker++
          this.uploadedFiles.set(uploadId, { name: event.name, size: event.size, status: 'done' })
        })
        .catch((e) => {
          console.error('Error uploading file ', event.name, ': ', e)
          this.uploadedFilesChangeTracker++
          this.uploadedFiles.set(uploadId, { name: event.name, size: event.size, status: 'error' })
        })
    },

    $_ocUploadingFileMessage(fileName) {
      const translated = this.$gettext('Uploading file "%{fileName}"')
      return this.$gettextInterpolate(translated, { fileName }, true)
    }
  }
}
</script>

<style lang="scss">
#files-drop-container {
  .oc-files-drop-drag-area {
    background: transparent;
    border: 1px dashed var(--oc-color-input-border);
    padding: var(--oc-space-xlarge);
  }
}
</style>
