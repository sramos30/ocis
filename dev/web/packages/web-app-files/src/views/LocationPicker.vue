<template>
  <main id="files-location-picker" class="oc-flex oc-height-1-1">
    <div tabindex="-1" class="files-list-wrapper oc-width-expand">
      <div id="files-location-picker-app-bar" class="oc-p-s">
        <h1 class="location-picker-selection-info oc-mb" v-text="title" />
        <p
          class="oc-text-muted oc-text-meta"
          data-testid="location-picker-current-hint"
          v-text="currentHint"
        />
        <hr class="oc-mt-rm" />
        <oc-breadcrumb :items="breadcrumbs" class="oc-mb-s" />
        <oc-grid gutter="small" flex class="oc-flex-middle">
          <div>
            <oc-button
              id="location-picker-btn-cancel"
              size="small"
              @click="leaveLocationPicker(originalLocation)"
            >
              <translate>Cancel</translate>
            </oc-button>
          </div>
          <div>
            <oc-button
              id="location-picker-btn-confirm"
              variation="primary"
              appearance="filled"
              size="small"
              :disabled="!canConfirm"
              @click="confirmAction"
            >
              <span v-text="confirmBtnText" />
            </oc-button>
          </div>
        </oc-grid>
      </div>
      <div id="files-view">
        <app-loading-spinner v-if="loadResourcesTask.isRunning" />
        <template v-else>
          <no-content-message
            v-if="isEmpty"
            id="files-location-picker-empty"
            class="files-empty"
            icon="folder"
          >
            <template #message>
              <span v-translate>There are no resources in this folder.</span>
            </template>
          </no-content-message>
          <resource-table
            v-else
            id="files-location-picker-table"
            class="files-table"
            :are-thumbnails-displayed="false"
            :resources="paginatedResources"
            :disabled="disabledResources"
            :target-route="targetRoute"
            :has-actions="false"
            :is-selectable="false"
            :header-position="fileListHeaderY"
            :sort-by="sortBy"
            :sort-dir="sortDir"
            @sort="handleSort"
          >
            <template #footer>
              <pagination :pages="paginationPages" :current-page="paginationPage" />
              <list-info
                v-if="paginatedResources.length > 0"
                class="oc-width-1-1 oc-my-s"
                :files="totalFilesCount.files"
                :folders="totalFilesCount.folders"
                :size="totalFilesSize"
              />
            </template>
          </resource-table>
        </template>
      </div>
    </div>
  </main>
</template>

<script>
import { mapMutations, mapState, mapActions, mapGetters } from 'vuex'

import { basename, join } from 'path'
import ResourceTable from '../components/FilesList/ResourceTable.vue'
import { batchActions } from '../helpers/batchActions'
import { cloneStateObject } from '../helpers/store'
import MixinsGeneral from '../mixins'
import MixinFilesListFilter from '../mixins/filesListFilter'
import { useTask } from 'vue-concurrency'

import NoContentMessage from 'web-pkg/src/components/NoContentMessage.vue'
import AppLoadingSpinner from 'web-pkg/src/components/AppLoadingSpinner.vue'
import ListInfo from '../components/FilesList/ListInfo.vue'
import Pagination from '../components/FilesList/Pagination.vue'
import { DavProperties } from 'web-pkg/src/constants'
import { createLocationPublic, createLocationSpaces } from '../router'
import { buildWebDavFilesPath, buildWebDavSpacesPath } from '../helpers/resources'
import { useResourcesViewDefaults } from '../composables'
import { useCapabilitySpacesEnabled } from 'web-pkg/src/composables'

export default {
  metaInfo() {
    const title = `${this.title} - ${this.configuration.currentTheme.general.name}`

    return { title }
  },

  components: {
    ResourceTable,
    NoContentMessage,
    AppLoadingSpinner,
    ListInfo,
    Pagination
  },

  mixins: [MixinsGeneral, MixinFilesListFilter],

  setup() {
    const loadResourcesTask = useTask(function* (signal, ref, target) {
      ref.CLEAR_CURRENT_FILES_LIST()

      if (typeof target === 'object') {
        target = ref.target
      }

      const { context } = ref.$route.params

      let resources

      switch (context) {
        case 'public':
          resources = yield ref.$client.publicFiles.list(
            target,
            ref.publicLinkPassword,
            DavProperties.Default
          )
          break
        case 'space':
          resources = yield ref.$client.files.list(
            buildWebDavSpacesPath(ref.$route.query.storageId, target),
            1,
            DavProperties.Default
          )
          break
        default:
          resources = yield ref.$client.files.list(
            buildWebDavFilesPath(ref.user.id, target),
            1,
            DavProperties.Default
          )
      }

      ref.loadFiles({ currentFolder: resources[0], files: resources.slice(1) })
      ref.loadIndicators({
        client: ref.$client,
        currentFolder: ref.$route.params.item || '/'
      })
      ref.refreshFileListHeaderPosition()
    }).restartable()

    return {
      ...useResourcesViewDefaults({ loadResourcesTask }),
      hasSpaces: useCapabilitySpacesEnabled()
    }
  },

  data: () => ({
    originalLocation: ''
  }),

  computed: {
    ...mapState('Files', [
      'selectedResourcesForMove',
      'locationPickerTargetFolder',
      'currentFolder'
    ]),
    ...mapGetters('Files', ['publicLinkPassword', 'totalFilesCount', 'totalFilesSize']),
    ...mapGetters(['configuration']),
    ...mapGetters(['homeFolder']),
    ...mapState(['user']),

    title() {
      const personalRouteTitle = this.hasSpaces
        ? this.$gettext('Personal')
        : this.$gettext('All files')
      const target =
        basename(this.target) || personalRouteTitle
          ? this.$gettext('Personal')
          : this.$gettext('All files')
      const translated =
        this.currentAction === batchActions.move
          ? this.$gettext('Move into »%{ target }«')
          : this.$gettext('Copy into »%{ target }«')
      return this.$gettextInterpolate(translated, { target })
    },

    currentAction() {
      return this.$route.params.action
    },

    isPublicContext() {
      return this.$route.params.context === 'public'
    },

    target() {
      return this.$route.params.item || this.homeFolder
    },

    resources() {
      const resources = cloneStateObject(this.$route.query.resource)

      // In case there is only one resource, ensure that the return will still be an array
      if (typeof resources === 'string') {
        return [resources]
      }

      return resources
    },

    resourcesCount() {
      return this.resources.length
    },

    breadcrumbs() {
      const breadcrumbs = []
      const pathSegments = this.target.split('/').filter((item) => item !== '')

      if (this.isPublicContext) {
        const itemPath = join.apply(null, pathSegments.slice(0, 1))
        breadcrumbs.push(this.createBreadcrumbNode(0, this.$gettext('Public link'), itemPath))
        for (let i = 1; i < pathSegments.length; i++) {
          const itemPath = join.apply(null, pathSegments.slice(0, i + 1))
          breadcrumbs.push(this.createBreadcrumbNode(i + 1, pathSegments[i], itemPath))
        }
      } else {
        const personalRouteTitle = this.hasSpaces
          ? this.$gettext('Personal')
          : this.$gettext('All files')
        breadcrumbs.push(this.createBreadcrumbNode(0, personalRouteTitle, '/'))
        for (let i = 0; i < pathSegments.length; i++) {
          const itemPath = join.apply(null, pathSegments.slice(0, i + 1))
          breadcrumbs.push(this.createBreadcrumbNode(i + 1, pathSegments[i], itemPath))
        }
      }

      // the very last breadcrumb node is not supposed to be clickable
      delete breadcrumbs[breadcrumbs.length - 1].to

      return breadcrumbs
    },

    canConfirm() {
      return this.currentFolder && this.currentFolder.canCreate()
    },

    confirmBtnText() {
      switch (this.currentAction) {
        case batchActions.move:
          return this.$pgettext('Confirm action in the location picker for move', 'Move here')
        case batchActions.copy:
          return this.$pgettext('Confirm action in the location picker for copy', 'Paste here')
        default:
          return this.$gettext('Confirm')
      }
    },

    disabledResources() {
      const resources = cloneStateObject(this.paginatedResources)

      return resources
        .filter((resource) => resource.type !== 'folder' || this.resources.includes(resource.path))
        .map((resource) => resource.id)
    },

    isEmpty() {
      return this.paginatedResources.length < 1
    },

    targetRoute() {
      const route = {
        name: this.$route.name,
        query: {
          resource: this.resources
        },
        params: {
          action: this.currentAction
        }
      }

      if (this.$route.query.storageId) {
        route.query.storageId = this.$route.query.storageId
      }

      return route
    },

    currentHint() {
      if (this.currentAction === 'move') {
        return this.$gettext(
          `Navigate into the desired folder and move selected resources into it.
          You can navigate into a folder by clicking on its name.
          To navigate back, you can click on the breadcrumbs.
          Resources will be moved into the folder where you are currently located.`
        )
      }

      return this.$gettext(
        `Navigate into the desired folder and copy selected resources into it.
        You can navigate into a folder by clicking on its name.
        To navigate back, you can click on the breadcrumbs.
        Resources will be copied into the folder where you are currently located.`
      )
    }
  },

  watch: {
    $route: {
      handler: function (to, from) {
        const sameRoute = to.name === from?.name
        const sameItem = to.params?.item === from?.params?.item
        if (!sameRoute || !sameItem) {
          this.loadResourcesTask.perform(this, this.$route)
        }
      },
      immediate: true
    }
  },

  created() {
    this.originalLocation = this.target
  },

  methods: {
    ...mapMutations('Files', ['CLEAR_CURRENT_FILES_LIST']),
    ...mapActions('Files', ['loadFiles', 'loadIndicators']),
    ...mapActions(['showMessage']),

    createBreadcrumbNode(index, text, itemPath) {
      return {
        index,
        text,
        to: {
          name: this.$route.name,
          params: {
            action: this.currentAction,
            item: itemPath
          },
          query: {
            resource: this.resources
          }
        }
      }
    },

    leaveLocationPicker(target) {
      switch (this.$route.params.context) {
        case 'public':
          this.$router.push(
            createLocationPublic('files-public-files', { params: { item: target } })
          )
          break
        case 'space':
          this.$router.push(
            createLocationSpaces('files-spaces-project', {
              params: { storageId: this.$route.query.storageId, item: target || '/' }
            })
          )
          break
        default:
          this.$router.push(
            createLocationSpaces('files-spaces-personal-home', { params: { item: target || '/' } })
          )
      }
    },

    isRowDisabled(resource) {
      if (resource.type !== 'folder' || !resource.canCreate()) {
        return true
      }
      return this.resources.some((item) => item === resource.path)
    },

    async confirmAction() {
      const errors = []
      let promise = null

      // Execute move or copy
      for (const resource of this.resources) {
        let targetPath
        let resourceName = basename(resource)
        const exists = this.paginatedResources.some((item) => {
          return basename(item.name) === resourceName
        })
        if (exists) {
          const message = this.$gettext('Resource with name %{name} already exists')
          errors.push({
            resource: resourceName,
            message: this.$gettextInterpolate(message, { name: resourceName }, true)
          })
          continue
        }

        switch (this.currentAction) {
          case batchActions.move: {
            switch (this.$route.params.context) {
              case 'public':
                targetPath = `${this.target || '/'}`
                targetPath += `/${resourceName}`
                promise = this.$client.publicFiles.move(
                  resource,
                  targetPath,
                  this.publicLinkPassword
                )
                break
              case 'space':
                targetPath = buildWebDavSpacesPath(this.$route.query.storageId, this.target || '/')
                targetPath += `/${resourceName}`
                promise = this.$client.files.move(
                  buildWebDavSpacesPath(this.$route.query.storageId, resource),
                  targetPath
                )
                break
              default:
                targetPath = buildWebDavFilesPath(this.user.id, this.target || '/')
                targetPath += `/${resourceName}`
                promise = this.$client.files.move(
                  buildWebDavFilesPath(this.user.id, resource),
                  targetPath
                )
            }
            break
          }
          case batchActions.copy: {
            switch (this.$route.params.context) {
              case 'public':
                targetPath = `${this.target || '/'}`
                targetPath += `/${resourceName}`

                promise = this.$client.publicFiles.copy(
                  resource,
                  targetPath,
                  this.publicLinkPassword
                )
                break
              case 'space':
                targetPath = buildWebDavSpacesPath(this.$route.query.storageId, this.target || '/')
                resourceName = basename(resource)
                targetPath += `/${resourceName}`
                promise = this.$client.files.copy(
                  buildWebDavSpacesPath(this.$route.query.storageId, resource),
                  targetPath
                )
                break
              default:
                targetPath = buildWebDavFilesPath(this.user.id, this.target || '/')
                targetPath += `/${resourceName}`
                promise = this.$client.files.copy(
                  buildWebDavFilesPath(this.user.id, resource),
                  targetPath
                )
            }
            break
          }
          default:
            return
        }

        await promise.catch((error) => {
          error.resource = resourceName
          errors.push(error)
        })
      }

      // Leave location picker if everything was successful
      if (errors.length === 0) {
        this.leaveLocationPicker(this.target)
        return
      }

      // Display error messages
      let title = ''
      if (errors.length === 1) {
        switch (this.currentAction) {
          case batchActions.move: {
            title = this.$gettext('Failed to move "%{resourceName}"')
            break
          }
          case batchActions.copy: {
            title = this.$gettext('Failed to copy "%{resourceName}"')
            break
          }
          default:
            return
        }

        this.showMessage({
          title: this.$gettextInterpolate(title, { resourceName: errors[0].resource }, true),
          status: 'danger'
        })

        return
      }

      switch (this.currentAction) {
        case batchActions.move: {
          title = this.$gettext('Failed to move %{count} resources')
          break
        }
        case batchActions.copy: {
          title = this.$gettext('Failed to copy %{count} resources')
          break
        }
        default:
          return
      }

      this.showMessage({
        title: this.$gettextInterpolate(title, { count: errors.length }),
        status: 'danger'
      })
    }
  }
}
</script>

<style lang="scss" scoped>
#files-location-picker {
  height: 100%;
  max-height: 100%;
  overflow-y: hidden;

  &-app-bar {
    position: sticky;
    top: 0;
    height: auto;
    z-index: 1;
    grid-area: header;
    background-color: var(--oc-color-background-default);
    box-sizing: border-box;
  }
}

.files-list-wrapper {
  height: 100%;
  max-height: 100%;
  overflow-y: auto;
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: max-content 1fr;
  gap: 0 0;
  grid-template-areas:
    'header'
    'main';

  &:focus {
    outline: none;
  }
}

#files-view {
  grid-area: main;
}

#files-location-picker-table ::deep tr {
  td:first-of-type {
    padding-left: 30px;
  }
}
</style>
