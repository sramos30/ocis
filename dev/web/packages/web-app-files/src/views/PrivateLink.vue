<template>
  <div class="oc-height-1-1 oc-link-resolve">
    <h1 class="oc-invisible-sr">{{ pageTitle }}</h1>
    <div class="oc-card oc-border oc-rounded oc-position-center oc-text-center oc-width-large">
      <template v-if="loading">
        <div class="oc-card-header">
          <h2 key="private-link-loading">
            <translate>Resolving private linkâ€¦</translate>
          </h2>
        </div>
        <div class="oc-card-body">
          <oc-spinner :aria-hidden="true" />
        </div>
      </template>
      <template v-else-if="errorMessage">
        <div class="oc-card-header oc-link-resolve-error-title">
          <h2 key="private-link-error">
            <translate>An error occurred while resolving the private link</translate>
          </h2>
        </div>
        <div class="oc-card-body oc-link-resolve-error-message">
          <p class="oc-text-lead">{{ errorMessage }}</p>
        </div>
      </template>
      <div class="oc-card-footer">
        <p>
          {{ configuration.currentTheme.general.slogan }}
        </p>
      </div>
    </div>
  </div>
</template>

<script>
import path from 'path'
import { mapGetters } from 'vuex'
import { buildResource, buildWebDavFilesPath } from '../helpers/resources'
import { createLocationSpaces } from '../router'
import { DavProperties } from 'web-pkg/src/constants'

export default {
  data() {
    return {
      loading: true,
      errorMessage: null
    }
  },
  computed: {
    ...mapGetters(['configuration', 'user']),

    pageTitle() {
      return this.$gettext(this.$route.meta.title)
    }
  },
  async mounted() {
    // query server to translate fileId to real path
    this.loading = true
    try {
      let resourcePath = await this.$client.files.getPathForFileId(this.$route.params.fileId)
      resourcePath = buildWebDavFilesPath(this.user.id, resourcePath)
      let resource = await this.$client.files.fileInfo(resourcePath, DavProperties.Default)
      resource = buildResource(resource)

      const params = {}
      const query = {}
      if (resource.isFolder) {
        // if folder: route directly into it
        params.item = resource.path || ''
      } else {
        // if file: route into parent and highlight file
        params.item = path.dirname(resource.path)
        query.scrollTo = resource.name
      }
      this.$router.push(
        createLocationSpaces('files-spaces-personal-home', {
          params,
          query
        })
      )
    } catch (error) {
      console.error(error)
      this.errorMessage = error
    } finally {
      this.loading = false
    }
  }
}
</script>

<style lang="scss">
.oc-link-resolve {
  .oc-card-header h2,
  .oc-card-footer p {
    margin: 0;
  }
}
</style>
