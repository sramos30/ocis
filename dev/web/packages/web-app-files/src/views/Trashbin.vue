<template>
  <trash-bin :breadcrumbs="breadcrumbs" />
</template>

<script>
import { bus } from 'web-pkg/src/instance'
import { useCapabilitySpacesEnabled } from 'web-pkg/src/composables'

import TrashBin from '../components/TrashBin.vue'

export default {
  components: { TrashBin },
  setup() {
    return {
      hasSpaces: useCapabilitySpacesEnabled()
    }
  },
  computed: {
    breadcrumbs() {
      const personalRouteName = this.hasSpaces
        ? this.$gettext('Personal')
        : this.$gettext('All files')

      return [
        {
          text: this.$gettext('Deleted files'),
          to: '/files/trash'
        },
        {
          text: personalRouteName,
          onClick: () => bus.publish('app.files.list.load')
        }
      ]
    }
  }
}
</script>
