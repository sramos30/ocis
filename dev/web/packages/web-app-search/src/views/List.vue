<template>
  <div class="oc-width-1-1">
    <div
      v-if="$asyncComputed.searchResults.updating"
      class="oc-flex oc-flex-middle oc-flex-center oc-height-1-1 oc-width-1-1"
    >
      <oc-spinner size="large" :aria-hidden="true" aria-label="" />
    </div>
    <template v-else>
      <component :is="listSearch.component" :search-results="searchResults" />
    </template>
  </div>
</template>

<script>
import { providerStore } from '../service'

export default {
  data() {
    const { provider: providerId } = this.$route.query
    const { listSearch } = providerStore.availableProviders.find(
      (provider) => provider.id === providerId
    )
    // abort and return if no provider is found
    return {
      listSearch
    }
  },

  asyncComputed: {
    searchResults: {
      get() {
        return this.listSearch.search(this.$route.query.term)
      },
      watch: ['$route.query.term', '$route.query.provider']
    }
  }
}
</script>

<style lang="scss">
#search-view {
  padding: 25px 10px 10px 10px;
}
</style>
