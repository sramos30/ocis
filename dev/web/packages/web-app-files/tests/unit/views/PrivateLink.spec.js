import { shallowMount } from '@vue/test-utils'
import { getRouter, getStore, localVue } from './views.setup'
import PrivateLink from '@files/src/views/PrivateLink.vue'
import fileFixtures from '../../../../../__fixtures__/files'

localVue.prototype.$client.files = {
  getPathForFileId: jest.fn(() => Promise.resolve('/lorem.txt'))
}

const theme = {
  general: { slogan: 'some-slogan' }
}

const $route = {
  params: {
    fileId: '2147491323'
  },
  meta: {
    title: 'Resolving private link'
  }
}

const selectors = {
  pageTitle: 'h1.oc-invisible-sr',
  loader: '.oc-card-body',
  errorTitle: '.oc-link-resolve-error-title'
}

describe('PrivateLink view', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  describe('when the page has loaded successfully', () => {
    let wrapper
    beforeEach(() => {
      wrapper = getShallowWrapper()
    })

    it('should display the page title', () => {
      expect(wrapper.find(selectors.pageTitle)).toMatchSnapshot()
    })
    it('should resolve the provided file id to a path', () => {
      expect(wrapper.vm.$client.files.getPathForFileId).toHaveBeenCalledTimes(1)
    })
  })

  describe('when the view is still loading', () => {
    it('should display the loading text with the spinner', () => {
      const wrapper = getShallowWrapper(true)

      expect(wrapper.find(selectors.loader)).toMatchSnapshot()
    })
  })

  describe('when there was an error', () => {
    it('should display the error message', async () => {
      jest.spyOn(console, 'error').mockImplementation(() => {})
      const wrapper = getShallowWrapper(
        false,
        jest.fn(() => Promise.reject(Error('some error')))
      )

      await new Promise((resolve) => {
        setTimeout(() => {
          expect(wrapper.find(selectors.errorTitle)).toMatchSnapshot()
          resolve()
        }, 1)
      })
    })
  })

  describe('when the view has finished loading and there was no error', () => {
    it('should not display the loading text and the error message', () => {
      const wrapper = getShallowWrapper()

      expect(wrapper).toMatchSnapshot()
    })
  })
})

function getShallowWrapper(loading = false, getPathForFileIdMock = jest.fn()) {
  return shallowMount(PrivateLink, {
    localVue,
    store: createStore(),
    mocks: {
      $route,
      $router: getRouter({}),
      $client: {
        files: {
          fileInfo: jest.fn().mockImplementation(() => Promise.resolve(fileFixtures['/'][4])),
          getPathForFileId: getPathForFileIdMock
        }
      }
    },
    data() {
      return {
        loading
      }
    }
  })
}

function createStore() {
  return getStore({
    slogan: theme.general.slogan,
    user: { id: 1 }
  })
}
