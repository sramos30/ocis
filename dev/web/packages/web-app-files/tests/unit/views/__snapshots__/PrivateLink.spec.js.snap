// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateLink view when the page has loaded successfully should display the page title 1`] = `<h1 class="oc-invisible-sr">Resolving private link</h1>`;

exports[`PrivateLink view when the view has finished loading and there was no error should not display the loading text and the error message 1`] = `
<div class="oc-height-1-1 oc-link-resolve">
  <h1 class="oc-invisible-sr">Resolving private link</h1>
  <div class="oc-card oc-border oc-rounded oc-position-center oc-text-center oc-width-large">
    <!---->
    <div class="oc-card-footer">
      <p>
        some-slogan
      </p>
    </div>
  </div>
</div>
`;

exports[`PrivateLink view when the view is still loading should display the loading text with the spinner 1`] = `
<div class="oc-card-body">
  <oc-spinner-stub arialabel="" size="medium" aria-hidden="true"></oc-spinner-stub>
</div>
`;

exports[`PrivateLink view when there was an error should display the error message 1`] = `
<div class="oc-card-header oc-link-resolve-error-title">
  <h2>
    <translate-stub tag="span">An error occurred while resolving the private link</translate-stub>
  </h2>
</div>
`;
